File: phonedirectory.go
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
// assign available numbers and check or release them
create map with available numbers (can use set too)
create queue to get next number
to get available numbers pop from q and set available false in map
return the number

to check return if its present or not in map

to release add into q and set true in map
*/
package main

import "fmt"

type PhoneDirectory struct {
	m map[int]bool
	q []int
}

func (p *PhoneDirectory) get() int {
	if len((*p).q) > 0 {
		n := (*p).q[0]
		(*p).q = (*p).q[1:]
		(*p).m[n] = false
		return n
	}
	return -1
}

func (p *PhoneDirectory) check(n int) bool {
	return (*p).m[n]
}

func (p *PhoneDirectory) release(n int) {
	(*p).q = append((*p).q, n)
	(*p).m[n] = true
}

func MainPhoneDirectory() {
	pd := &PhoneDirectory{
		m: make(map[int]bool),
		q: []int{},
	}

	for i := 0; i < 10; i++ {
		(*pd).m[i] = true
		(*pd).q = append((*pd).q, i)
	}

	fmt.Println((*pd).get())    // 0
	fmt.Println((*pd).get())    // 1
	fmt.Println((*pd).check(2)) // true
	fmt.Println((*pd).get())    // 2
	fmt.Println((*pd).check(2)) // false
	(*pd).release(2)            // nil
	fmt.Println((*pd).check(2)) // true

}

File: Phone Directory.java
// Time Complexity : O(1)
// Space Complexity : O(maxNumbers)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach
// We will maintain a hashset and a queue to store the available numbers. We will put it in hashset to make the search optimal and we put it in the queue to get the available numbers by popping them. When the queue is empty, it means no element is available. Whenever we want a number, we just pop the number from the queue. To release a number i.e to make a number available in the queue, we add it to the set and the queue.

class PhoneDirectory 
{
    HashSet <Integer> set;
    Queue <Integer> q;

    public PhoneDirectory(int maxNumbers) 
    {
        this.set=new HashSet<>();
        this.q=new LinkedList<>();

        for(int i=0;i<maxNumbers;i++)
        {
            set.add(i); //add the number to the set
            q.add(i);   //add the number to the queue
        }
    }
    
    public int get() 
    {
        if(q.isEmpty()) return -1;  //if no number is available return -1
        int re=q.poll();    //else return the popped number from the queue
        set.remove(re);     //remove the same number from the set as well
        return re;
    }
    
    public boolean check(int number) 
    {
        return set.contains(number);    //if the number is present in the set, return true else false
    }
    
    public void release(int number) 
    {
        //if the number is not present in the set, then add it 
        if (!set.contains(number))
        {
            set.add(number);
            q.add(number);
        }
    }
}
